File: picasso-commons/src/com/hendraanggrian/picasso/commons/transformation/MaskTransformer.java
package com.hendraanggrian.picasso.commons.transformation;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.lang.ref.WeakReference;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
class MaskTransformer extends Transformer {

    @Nullable private static WeakReference<Paint> paint;

    @NonNull private final Drawable mask;

    MaskTransformer(@NonNull Drawable mask) {
        this.mask = mask;
    }

    @NonNull
    @Override
    protected Bitmap transform(@NonNull Bitmap source, boolean shouldRecycle) {
        int width = source.getWidth();
        int height = source.getHeight();
        Bitmap result = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);

        Canvas canvas = new Canvas(result);
        mask.setBounds(0, 0, width, height);
        mask.draw(canvas);
        canvas.drawBitmap(source, 0, 0, getPaint());

        if (shouldRecycle) {
            source.recycle();
        }
        return result;
    }

    @NonNull
    @Override
    protected Bundle keyBundle() {
        Bundle bundle = new Bundle();
        bundle.putString(KEY_NAME, "MaskTransformer");
        return bundle;
    }

    @NonNull
    private Paint getPaint() {
        Paint instance;
        if (paint != null) {
            instance = paint.get();
            if (instance != null) {
                return instance;
            }
        }
        instance = new Paint();
        instance.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_IN));
        paint = new WeakReference<>(instance);
        return instance;
    }
}

File: picasso-commons/src/com/hendraanggrian/picasso/commons/transformation/MaskTransformer.kt
package com.hendraanggrian.picasso.commons.transformation

import android.graphics.*
import android.graphics.drawable.Drawable
import android.os.Bundle
import java.lang.ref.WeakReference

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
internal class MaskTransformer(private val mask: Drawable) : Transformer() {

    companion object {
        private var paint: WeakReference<Paint>? = null
    }

    override fun transform(source: Bitmap, shouldRecycle: Boolean): Bitmap {
        val width = source.width
        val height = source.height
        val result = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)

        val canvas = Canvas(result)
        mask.setBounds(0, 0, width, height)
        mask.draw(canvas)
        canvas.drawBitmap(source, 0f, 0f, getPaint())

        if (shouldRecycle) {
            source.recycle()
        }
        return result
    }

    override fun keyBundle(): Bundle {
        val bundle = Bundle()
        bundle.putString(Transformer.KEY_NAME, "MaskTransformer")
        return bundle
    }

    private fun getPaint(): Paint {
        var instance: Paint?
        if (paint != null) {
            instance = paint!!.get()
            if (instance != null) {
                return instance
            }
        }
        instance = Paint()
        instance.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_IN)
        paint = WeakReference(instance)
        return instance
    }
}